### DISCUSSION TITLE
Fighting with Push Error: PreconditionFailed

### DISCUSSION THREAD

**Discussion by oliverw** [Azure/azure-mobile-apps#638] [Q&A]
I know this is hard to diagnose without seeing the actual code, so this is more of a brainstorming thread. 

My project is some kind of a task management app where a task pretty much boils down to a Title and Status (Created, Completed). The app is Rx based and since I'm utilizing the excellent [DynamicData](https://github.com/reactivemarbles/DynamicData) library, it was pretty easy to implement a queue that is fed by a handler listening to changes of my collection of tasks and changes of properties on tasks contained in the collection. The handler servicing the queue then updates the Datasync IOfflineTable<T> according to the observed change:

```c#
private async Task ServiceDatasyncActionQueue(CancellationToken ct)
{
    while(!ct.IsCancellationRequested)
    {
        try
        {
            var entityAction = await actionQueue.ReceiveAsync(ct);

            switch (entityAction.Entity)
            {
                case Model.Task task:
                    await PerformEntityAction(task, Tasks, entityAction.Action);
                    break;
            }
        }

        catch (Exception ex)
        {
            this.Log().Error(ex);
        }
    }
}

private async Task PerformEntityAction<T>(T entity, IOfflineTable<T> table, EntityActionType action)
    where T: Model.IDatasyncEntity
{
    Contract.RequiresNonNull(entity);
    Contract.RequiresNonNull(table);

    switch (action)
    {
        case EntityActionType.Create:
            await table.InsertItemAsync(entity);
            break;

        case EntityActionType.Update:
            await table.ReplaceItemAsync(entity);
            break;

        case EntityActionType.Delete:
            await table.DeleteItemAsync(entity);
            break;
    }

    RequestTableSync(false);
}
```

The `RequestTableSync(false)` call above then asynchronously requests the a backend sync operation. The backend sync boils down to this code:

```c#
private async Task SyncTablesAsync(SyncRequest request, CancellationToken ct)
{
    try
    {
        await SyncTableAsync(Tasks, ct);
    }

    catch (Exception ex)
    {
        this.Log().Error(ex);
    }
}

private async Task SyncTableAsync<T>(IOfflineTable<T> table, CancellationToken ct)
    where T: Model.ReactiveDatasyncEntity
{
    try
    {
        await table.PushItemsAsync(ct);
        await table.PullItemsAsync("", ct);

        var deletedItemsQuery = table.CreateQuery().Where(x => x.Deleted);

        await table.PurgeItemsAsync(deletedItemsQuery, new PurgeOptions
        {
            DiscardPendingOperations = true
        }, ct);
    }

    catch (PushFailedException ex)
    {
        this.Log().Error($"{SyncLogPrefix} Encountered {ex.PushResult?.Errors?.Count ?? 0} push error(s) with status {ex.PushResult?.Status} during sync: {string.Join("\n", ex.PushResult?.Errors?.Select(x => x.RawResult) ?? Enumerable.Empty<string>())}");

        if (ex.PushResult?.Errors != null)
        {
            foreach (var error in ex.PushResult.Errors)
            {
                if (error.OperationKind == TableOperationKind.Update && error.Result != null)
                {
                    // Update failed, reverting to server's copy
                    await error.CancelAndUpdateItemAsync(error.Result, ct);

                }
                else
                {
                    // Discard local change
                    await error.CancelAndDiscardItemAsync(ct);
                }

                this.Log().Error($"{SyncLogPrefix} Error executing sync operation. Item: {error.TableName} ({(error.Item["id"] ?? "<null item id>")}). Operation discarded.");
            }
        }
    }
}
```

I'm able to trigger the `PushFailedException` above pretty often by touching an entity again in rapid succession. The reason is always `PreconditionFailed` which probably is a mismatch of the `Version` property. I'm having trouble wrapping my head around all the moving parts 😄

**Comment by adrianhall** [comment]
PreconditionFailed only occurs on a version mismatch.  I can almost guarantee you are bumping into an issue where you an initiating a Push during a Pull of the same table.  What you need to do is implement collapsible queuing through the RequestTableSync call.   In essence, create a queuing system for handling synchronization requests.  Instead of directly requesting the synchronization, push it onto the queue (a pending request) only if there isn't already a pending request for the same table.  Then your queue processor can take things off the queue one at a time and do the appropriate synchronization.

**Reply by oliverw** [reply]
> PreconditionFailed only occurs on a version mismatch. I can almost guarantee you are bumping into an issue where you an initiating a Push during a Pull of the same table. What you need to do is implement collapsible queuing through the RequestTableSync call. In essence, create a queuing system for handling synchronization requests. Instead of directly requesting the synchronization, push it onto the queue (a pending request) only if there isn't already a pending request for the same table. Then your queue processor can take things off the queue one at a time and do the appropriate synchronization.

Nevermind. Have to think it over.

**Comment by oliverw** [comment]
Rewrote the queuing logic using an `ActionBlock` with a `BoundedCapacity` of 1. Looking good so far. TPL Dataflow is such a godsent for stuff like this.

**Comment by oliverw** [comment]
What could actually go wrong if you have exactly one thread writing to the offline store and another thread performing Push/Pulls? Is that already a race condition?

My architecture currently allows for ..

- Multiple updates to an offline store record before a push/pull happens to sync with the backend
- There can be updates to the offline store while a Push/Pull is in flight

Just to make sure that I'm not doing something stupid/unsupported here.

**Reply by adrianhall** [reply]
Generally, there are no problems with having regular updates + pushes happening at the same time.

The corner case:
1. You create a new record (this enqueues a POST entry in the operations queue)
2. You start pushing a large update.
3. During the push, you DELETE the new record (this deletes the entry in the operations queue)
4. The operations queue entry has already been read, and it tries to execute the POST, but the operations queue entry no longer exists, resulting in a failure.

The failure is handled gracefully, but you'll get a push failure with a null reference exception.

**Reply by oliverw** [reply]
Hmm. Still able to provoke `PreconditionFailed` Push-Errors from time to time.

**Reply by adrianhall** [reply]
I'd like to understand the case(s) where this happens unexpectedly so we can correct.  Without concrete reproductions, it's going to be impossible, though.
